### DISCUSSION TITLE
How do I install Language packs?

### DISCUSSION THREAD

**Discussion by vnetonline** [oqtane/oqtane.framework#2700] [General]
I have started to play around with Oqtane but can't figure out how I can install a language pack can someone please guide me?

When I go to language Management I don't see the languages published on the market place

![image](https://user-images.githubusercontent.com/15242923/228128121-7c015a94-7321-4817-8960-cec16d4fb6e1.png)

**Comment by sbwalker** [comment]
Language Management allows you to specify the languages which are enabled in your site. The expected flow is that a site owner would know which languages need to be enabled in their site... and they would use the Add Language option to add languages (note that you need to add a language BEFORE you install a translation). 

In Add Language, the Translated? option was intended to filter the list of languages to those which have translations already installed in your installation - it has nothing to do with what is available in the Marketplace. This option is confusing so I will remove it from the UI.

Once a language has been added it will be displayed in the list in the languages. If a translation exists in the Marketplace, an Install button should be displayed (note that there was a bug which was affecting this behavior that has now been fixed in #2703).

![image](https://user-images.githubusercontent.com/4840590/228354956-842622a7-1ec8-4688-829d-4b28e6a41cdb.png)

For languages where a translation has been previously installed, the version number of the translation will be displayed. If a newer version of the translation becomes available in the Marketplace, an Upgrade button will be displayed.

**Comment by vnetonline** [comment]
Thanks Shaun 

I can confirm that I can now add the language packs

![image](https://user-images.githubusercontent.com/15242923/228401620-3e38923e-a863-4d6a-be99-cd18d82ddc14.png)

* Once added i can see the download buttons 

![image](https://user-images.githubusercontent.com/15242923/228401696-c08ee28f-b4ae-4a53-bb32-c4b7c9ebaa4d.png)

- Once I download I can see the `Review License Treams`

![image](https://user-images.githubusercontent.com/15242923/228401822-13280c71-22b3-4524-8135-e533f2dfcd0e.png)

- Once I accept and click on Install I get the below error 

![image](https://user-images.githubusercontent.com/15242923/228403647-99b76443-7267-4093-a135-34d75d9ae1af.png)

- I think the Urdu (Pakistan) NuGet package is corrupt, so this should be handled by a try-catch block and write a message in the log and delete the nupkg?

**Comment by sbwalker** [comment]
@isaeed there seems to be a problem with your translation package

Oqtane Error: Installing Package Oqtane.Framework.ur-PK.3.4.0.nupkg - System.IO.InvalidDataException: End of Central Directory record could not be found.
   at System.IO.Compression.ZipArchive.ReadEndOfCentralDirectory()
   at System.IO.Compression.ZipArchive..ctor(Stream stream, ZipArchiveMode mode, Boolean leaveOpen, Encoding entryNameEncoding)
   at System.IO.Compression.ZipFile.Open(String archiveFileName, ZipArchiveMode mode, Encoding entryNameEncoding)
   at System.IO.Compression.ZipFile.OpenRead(String archiveFileName)
   at Oqtane.Infrastructure.InstallationManager.InstallPackages(String webRootPath, String contentRootPath) in Oqtane.Server\Infrastructure\InstallationManager.cs:line 85

**Comment by ijaz-saeed** [comment]
@sbwalker thanks to point it out, but I just downloaded the package and renamed it as .zip and window explorer is able to unzip it. so, not sure what is the actual problem. will try to investigate more. 

<img width="683" alt="image" src="https://user-images.githubusercontent.com/21291112/228577578-6bbc308f-2a07-4a72-bf17-e04174ec8f3d.png">

**Reply by sbwalker** [reply]
@ijaz-saeed are you using the standard "nuget.exe pack *.nuspec" to create your Nuget packages - or are you using a third party ZIP utility?

**Reply by ijaz-saeed** [reply]
hi, @sbwalker yes using nuget.exe and here is the complete command , 

dotnet build -c Release ..\..\Oqtane.sln
mkdir ..\..\Oqtane.Server\bin\Debug\net6.0\ur-PK
copy ..\..\Oqtane.Server\bin\Release\net6.0\ur-PK\Oqtane.Client.resources.dll ..\..\Oqtane.Server\bin\Debug\net6.0\ur-PK\Oqtane.Client.resources.dll
..\..\Oqtane.Package\nuget.exe pack Oqtane.Framework.ur-PK.nuspec
pause

**Reply by ijaz-saeed** [reply]
hi @sbwalker I have found the issue, the package download Url was wrong. I have corrected it and seems to be working now.  

**_https://www.nuget.org/packages/Oqtane.Framework.ur-PK -- Wrong_** 

**_https://www.nuget.org/api/v2/package/Oqtane.Framework.ur-PK/3.4.0 -- Valid_** 

so, instead of using download Url, I was wronging using main Package Url.

**Reply by sbwalker** [reply]
@ijaz-saeed thank you for identifying the cause of the issue

**Comment by vnetonline** [comment]
Thanks @sbwalker 

Shouldn't there be an error on the  UI if a package fails, I would like to send a pull request to more elegantly handle if an installation of a package fails.

The proposed solution is as follows 

1. The `InstallPackages` method should return a `List<string>` (errorList) detailing the package name which failed 
2. This `errorList` should be passed up the chain to the PackageController and log the `errorList` in the EventLog.
3. The same `errorList` list will be passed to the PackageService and to the UI 
4. Then we can display an error on the UI 

This way we can log the error and the user can get a visual that a package failed to install like this

![image](https://user-images.githubusercontent.com/15242923/228698370-5a99eb87-a0a2-487c-ac5b-619b1d32dec9.png)

When you then click on view details you will be able to see this 

![image](https://user-images.githubusercontent.com/15242923/228698480-34c9c244-fbc9-4d49-99ff-aab517d82020.png)

**Reply by sbwalker** [reply]
It's a bit more challenging because there are multiple scenarios which execute the InstallPackages logic. There is a non-interactive scenario during startup... and an interactive scenario when the framework is already running. As a result of the startup scenario, the InstallPackages method had to be declared as a static... which makes logging more difficult as you can't use standard DI. Also adding the logging service in this class caused issues in a clean install scenario as the database has not yet been created. So although I agree that logging would be useful, there are a variety of technical challenges which would need to be addressed.

**Reply by vnetonline** [reply]
We can at least do the logging for the controllers which are called by the services which the UI interact with ?

To resolve the issue of not having the logger in non interactive modes where installation is done we could implement a service locator pattern

Service Locator
Service locator pattern is another way of obtaining dependencies. 

```
Example:
public class PackageService
{
    private readonly IPackageRepository _packageRepository;
    private readonly ILogger<PackageService> _logger;
    public PackageService(IServiceProvider serviceProvider)
    {
        _packageRepository = serviceProvider
          .GetRequiredService<IPackageRepository>();
        _logger = serviceProvider
          .GetService<ILogger<PackageService>>() ??
            NullLogger<PackageService>.Instance;
    }
    public void Delete(int id)
    {
        _packageRepository.Delete(id);
        _logger.LogInformation($"Deleted a product with id = {id}");
    }
}
```

**Reply by sbwalker** [reply]
The Service Locator pattern is used in other areas of the framework already... we try to minimize its usage as it tends to be an anti-pattern. And it does not resolve the scenario of a clean install where the database does not yet exist. There is an alternate file logger for these types of scenarios (ie. scenarios where the database does not exist, there is no context, the tenant cannot be determined, etc.. ) - which is surfaced in the System Info area (as it is not specific to a site - it is related to the installation). This would seem to be the appropriate place to log package failures.

**Reply by sbwalker** [reply]
One other thing to keep in mind... the InstallPackages() method is only responsible for unzipping a Nuget package and copying files to the appropriate locations. At this stage the package is not actually "installed" - it is merely deployed. In order to complete the installation process, the framework run-time process needs to be restarted. This is required because DI service registration can only occur during startup in .NET Core. When the framework starts up it is able to identify the assemblies which need to be loaded and interrogated for various interfaces... such as the interface for EF Core Migrations, etc... It will execute these interfaces to complete the installation process - which is a non-interactive process. It is during this process that issues are more likely to arise (ie. failed DB migrations, etc...) and they are logged to the file logger. InstallPackages() will only fail if the Nuget package is corrupt - which seemed like a fairly unlikely scenario... until it happened with the Urdu language pack ;)

**Reply by vnetonline** [reply]
Okay makes sense 

So we can't even use the fileLogger in the static method if the package is corrupt?

But instead of using debug, can we pass the _fileLogger into the static method and log the error?

In a production environment if a package is corrupt and can't be unzipped the host will not know why the package is not installed just gets deleted.

**Reply by sbwalker** [reply]
@vnetonline yes I totally agree that there needs to be logging for this scenario - my comments above were simply explaining that there were some technical challenges and context to consider in doing so which are not immediately obvious. I simply added a basic try... catch in #2704 as I needed to get 3.4.2 released... but proper logging is needed.

**Reply by vnetonline** [reply]
@sbwalker as part of upgrade of Oqtane to 4.0 which will be on.Net 7 

Can we switch to 

using Serilog;
using Serilog.Events;

That way we can use it in static methods

Something like this in Program.cs

```
var loggerConfiguration = new LoggerConfiguration()
#if DEBUG
            .MinimumLevel.Debug()
#else
            .MinimumLevel.Information()
#endif
            .MinimumLevel.Override("Microsoft", LogEventLevel.Information)
            .MinimumLevel.Override("Microsoft.EntityFrameworkCore", LogEventLevel.Warning)
            .Enrich.FromLogContext()
            .WriteTo.Async(c => c.File("Logs/logs.txt"))
            .WriteTo.Async(c => c.Console());

        if (IsMigrateDatabase(args))
        {
            loggerConfiguration.MinimumLevel.Override("Microsoft", LogEventLevel.Warning);
        }

        Log.Logger = loggerConfiguration.CreateLogger();

        try
        {
            var builder = WebApplication.CreateBuilder(args);
            builder.Host.AddAppSettingsSecretsJson()
                .UseAutofac()
                .UseSerilog();
            if (IsMigrateDatabase(args))
            {
                builder.Services.AddDataMigrationEnvironment();
            }
            await builder.AddApplicationAsync<OqtaneModule>();
            var app = builder.Build();
            await app.InitializeApplicationAsync();

            if (IsMigrateDatabase(args))
            {
                await app.Services.GetRequiredService<MyProjectNameDbMigrationService>().MigrateAsync();
                return 0;
            }

            Log.Information("Starting Oqtane.Server");
            await app.RunAsync();
            return 0;
        }
        catch (Exception ex)
        {
            if (ex is HostAbortedException)
            {
                throw;
            }

            Log.Fatal(ex, "Oqtane.Server terminated unexpectedly!");
            return 1;
        }
        finally
        {
            Log.CloseAndFlush();
        }
```

**Reply by sbwalker** [reply]
@vnetonline Oqtane has an overarching Philosophy (https://www.oqtane.org/blog/!/20/oqtane-philosophy) which includes minimizing dependencies on third party libraries. 

I had investigated Serilog early in Oqtane's evolution but could not easily determine how to allow multi-tenant support on the server or Blazor support on the client - which is 95% of the logging required by Oqtane. So instead I developed a custom logger which supports the specific capabilities needed by the framework - and provides simple methods for module developers. The custom logger uses the same approach as Serilog for defining log messages with substitutable tokens. 

There is still 5% of logging which is needed for very specific scenarios (described earlier in this thread) which is where the file logger was introduced. 

I am not sure if I fully understand your question... but based on the code you included, it would not satisfy the needs of Oqtane - nor would it align with the philosophy.

**Comment by vnetonline** [comment]
Fair enough, I understand …. Let me know how I can help …. Happy to research how we can solve this which aligns with the philosophy. 

I would be happy to help with documentation. Let me know where you need the most help.

**Reply by sbwalker** [reply]
@vnetonline #2730 addresses the issue you identified with logging for package installation. The PR ended up being quite large because I took the opportunity to consolidate the behavior so that package installation occurs consistently for all scenarios in the framework. So there is no longer an interactive flow for extracting a package - all scenarios rely on the same non-interactive flow during Startup.

**Reply by vnetonline** [reply]
Thanks, for this @sbwalker brilliant effort really appreciate this project and have been looking for something like this done using a modern framework rather then DNN.

**Comment by vnetonline** [comment]
I can confirm that issue is now resolved because all scenarios rely on the same non-interactive flow during Startup.

![image](https://user-images.githubusercontent.com/15242923/230273047-f2f991e3-0f46-4c4d-9a91-c489a8fd750d.png)

The above shows the error in `Log` tab of the `System Info` section of `Admin Dashboard` if the package is curupted
